test/java/seedu/addressbook/parser/ParserTest.java

```diff
@@ -0,0 +1,302 @@
+        parseAndAssertIncorrectWithMessage(resultMessage, inputs);
+    }
+
+    @Test
+    public void addCommand_invalidPersonDataInArgs() {
+        final String invalidName = "[]\\[;]";
+        final String validName = Name.EXAMPLE;
+        final String invalidPhoneArg = "p/not__numbers";
+        final String validPhoneArg = "p/" + Phone.EXAMPLE;
+        final String invalidEmailArg = "e/notAnEmail123";
+        final String validEmailArg = "e/" + Email.EXAMPLE;
+        final String invalidTagArg = "t/invalid_-[.tag";
+
+        // address can be any string, so no invalid address
+        final String addCommandFormatString = "add $s $s $s a/" + Address.EXAMPLE;
+
+        // test each incorrect person data field argument individually
+        final String[] inputs = {
+                // invalid name
+                String.format(addCommandFormatString, invalidName, validPhoneArg, validEmailArg),
+                // invalid phone
+                String.format(addCommandFormatString, validName, invalidPhoneArg, validEmailArg),
+                // invalid email
+                String.format(addCommandFormatString, validName, validPhoneArg, invalidEmailArg),
+                // invalid tag
+                String.format(addCommandFormatString, validName, validPhoneArg, validEmailArg) + " " + invalidTagArg
+        };
+        for (String input : inputs) {
+            parseAndAssertCommandType(input, IncorrectCommand.class);
+        }
+    }
+
+    @Test
+    public void addCommand_validPersonData_parsedCorrectly() {
+        final Person testPerson = generateTestPerson();
+        final String input = convertPersonToAddCommandString(testPerson);
+        final AddPersonCommand result = parseAndAssertCommandType(input, AddPersonCommand.class);
+        assertEquals(result.getPerson(), testPerson);
+    }
+
+    @Test
+    public void addCommand_duplicateTags_merged() throws IllegalValueException {
+        final Person testPerson = generateTestPerson();
+        String input = convertPersonToAddCommandString(testPerson);
+        for (Tag tag : testPerson.getTags()) {
+            // create duplicates by doubling each tag
+            input += " t/" + tag.tagName;
+        }
+
+        final AddPersonCommand result = parseAndAssertCommandType(input, AddPersonCommand.class);
+        assertEquals(result.getPerson(), testPerson);
+    }
+
+    private static Person generateTestPerson() {
+        try {
+            return new Person(
+                new Name(Name.EXAMPLE),
+                new Phone(Phone.EXAMPLE, true),
+                new Email(Email.EXAMPLE, false),
+                new Address(Address.EXAMPLE, true),
+                new UniqueTagList(new Tag("tag1"), new Tag("tag2"), new Tag("tag3"))
+            );
+        } catch (IllegalValueException ive) {
+            throw new RuntimeException("test person data should be valid by definition");
+        }
```

Comment:
If this is not supposed to happen, just add a throws clause to the method (no need to catch and handle)